lib/IDE/CompletionInstance.cpp

```diff
@@ -0,0 +1,360 @@
+}
+
+namespace {
+/// Returns index number of \p D in \p Decls . If it's not found, returns ~0.
+template <typename Range>
+unsigned findIndexInRange(Decl *D, const Range &Decls) {
+  unsigned N = 0;
+  for (auto I = Decls.begin(), E = Decls.end(); I != E; ++I) {
+    if (*I == D)
+      return N;
+    ++N;
+  }
+  return ~0U;
+}
+
+/// Return the element at \p N in \p Decls .
+template <typename Range> Decl *getElementAt(const Range &Decls, unsigned N) {
+  auto I = Decls.begin();
+  std::advance(I, N);
+  return *I;
+}
+
+/// Find the equivalent \c DeclContext with \p DC from \p SF AST.
+/// This assumes the AST which contains \p DC has exact the same structure with
+/// \p SF.
+/// FIXME: This doesn't support IfConfigDecl blocks. If \p DC is in an inactive
+///        config block, this function probably returns false.
+static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
+                                                           SourceFile *SF) {
+  auto *newDC = DC;
+  // NOTE: Shortcut for DC->getParentSourceFile() == SF case is not needed
+  // because they should be always different.
+
+  // Get the index path in the current AST.
+  SmallVector<unsigned, 4> IndexStack;
+  do {
+    auto *D = newDC->getAsDecl();
+    auto *parentDC = newDC->getParent();
+    unsigned N;
+    if (auto parentSF = dyn_cast<SourceFile>(parentDC))
+      N = findIndexInRange(D, parentSF->Decls);
+    else if (auto parentIDC =
+                 dyn_cast<IterableDeclContext>(parentDC->getAsDecl()))
+      N = findIndexInRange(D, parentIDC->getMembers());
+    else
+      llvm_unreachable("invalid DC kind for finding equivalent DC (indexpath)");
+
+    // Not found in the decl context tree.
+    // FIXME: Probably DC is in an inactive #if block.
+    if (N == ~0U) {
+      return nullptr;
+    }
+
+    IndexStack.push_back(N);
+    newDC = parentDC;
+  } while (!newDC->isModuleScopeContext());
+
+  assert(isa<SourceFile>(newDC) && "DC should be in a SourceFile");
+
+  // Query the equivalent decl context from the base SourceFile using the index
+  // path.
+  newDC = SF;
+  do {
+    auto N = IndexStack.pop_back_val();
+    Decl *D;
+    if (auto parentSF = dyn_cast<SourceFile>(newDC))
+      D = getElementAt(parentSF->Decls, N);
+    else if (auto parentIDC = dyn_cast<IterableDeclContext>(newDC->getAsDecl()))
+      D = getElementAt(parentIDC->getMembers(), N);
+    else
+      llvm_unreachable("invalid DC kind for finding equivalent DC (query)");
+    newDC = dyn_cast<DeclContext>(D);
+  } while (!IndexStack.empty());
+
+  assert(newDC->getContextKind() == DC->getContextKind());
+
+  return newDC;
+}
+
+/// Calculate the hash value of the \c CompilerInvocation.
+/// This should take all options affecting completion results into account.
+/// If the hashes between multiple completion invocation are different, we
+/// cannot reuse the CompilerInstance.
+static llvm::hash_code calculateInvocationHash(const CompilerInvocation &Inv) {
```

Comment:
How does this compare to `getPCHHash()`?  Can we reuse some of that logic?

Reply:
`getPCHHash()` doesn't take `FrontendOptions` into account. That means it's not updated even when other files are added to the project.

> Can we reuse some of that logic?

Probably yes. But I haven't looked into it

Reply:
This function is now completely gone.